### ISSUE TITLE
Document editing not working behind HTTPS reverse proxy

### ISSUE THREAD

**Issue by zarbis** [ONLYOFFICE/Docker-CommunityServer#10]
In my setup I use two containers for Community and Document servers. Everything worked fine. My docker-compose.yml file is as follows:

```
version: '2'
services:

  community:
    restart: always
    image: onlyoffice/communityserver
    depends_on:
      - document
    networks:
      default:
      proxy:
    volumes:
      - ./CommunityServer/data:/var/www/onlyoffice/Data
      - ./CommunityServer/mysql:/var/lib/mysql
      - ./CommunityServer/logs:/var/log/onlyoffice
      - ./DocumentServer/data:/var/www/onlyoffice/DocumentServerData
    environment:
      VIRTUAL_HOST: office.mydomain.com
      LETSENCRYPT_HOST: office.mydomain.com
      LETSENCRYPT_EMAIL: admin@mydomain.com
      DOCUMENT_SERVER_PORT_80_TCP_ADDR: document
    stdin_open: true
    tty: true

  document:
    restart: always
    image: onlyoffice/documentserver
    volumes:
      - ./DocumentServer/data:/var/www/onlyoffice/Data
      - ./DocumentServer/logs:/var/log/onlyoffice
    stdin_open: true
    tty: true

networks:
  proxy:
    external:
      name: nginxproxy_backend
```

I use nginx as a reverse proxy for my services. Recently I've set up HTTPS for all my services with Let's Encrypt so config for each service looks like this:

```
upstream office.mydomain.com {
                ## Can be connect with "nginxproxy_backend" network
            # onlyoffice_community_1
            server 172.19.0.7:80;
}
server {
    server_name office.mydomain.com;
    listen 80 ;
    access_log /var/log/nginx/access.log vhost;
    return 301 https://$host$request_uri;
}
server {
    server_name office.mydomain.com;
    listen 443 ssl http2 ;
    access_log /var/log/nginx/access.log vhost;
    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA;
    ssl_prefer_server_ciphers on;
    ssl_session_timeout 5m;
    ssl_session_cache shared:SSL:50m;
    ssl_certificate /etc/nginx/certs/office.mydomain.com.crt;
    ssl_certificate_key /etc/nginx/certs/office.mydomain.com.key;
    ssl_dhparam /etc/nginx/certs/office.mydomain.com.dhparam.pem;
    add_header Strict-Transport-Security "max-age=31536000";
    include /etc/nginx/vhost.d/default;
    location / {
        proxy_pass http://office.mydomain.com;
    }
}
```

The problem I'm encountering is: 
1) open document for editing
2) new tab appears and editor interface loaded
3) content of document doesn't load, loading banner spins forever
4) in address bar there is warning about mixed content
5) if i click "disable protection for now" page reloads and i immediately get "Unknown error" banner

**Comment by agolybev** [comment]
You need add some http headers when proxying request
Try adding the following string into "location /" section
proxy_set_header Host $host;
proxy_set_header X-Forwarded-Proto $scheme;

**Comment by Zarbis** [comment]
I'm using jwilder/nginx-proxy for my containers. Aside of server blocks for each container it generates this part:
```
# If we receive X-Forwarded-Proto, pass it through; otherwise, pass along the
# scheme used to connect to this server
map $http_x_forwarded_proto $proxy_x_forwarded_proto {
  default $http_x_forwarded_proto;
  ''      $scheme;
}
# If we receive Upgrade, set Connection to "upgrade"; otherwise, delete any
# Connection header that may have been passed to this server
map $http_upgrade $proxy_connection {
  default upgrade;
  '' close;
}
gzip_types text/plain text/css application/javascript application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
log_format vhost '$host $remote_addr - $remote_user [$time_local] '
                 '"$request" $status $body_bytes_sent '
                 '"$http_referer" "$http_user_agent"';
access_log off;
# HTTP 1.1 support
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
```
Note first and last lines, seems like equivalent to me.

**Comment by TMHBOFH** [comment]
Halle  Zarbis,

did you found a solution for this problem ?
I have the same issue.

**Comment by Zarbis** [comment]
@TMHBOFH, I haven't found a solution that would keep my existing setup.
As much as I understand the problem lies between those facts:
1) HTTPS is terminated on reverse proxy, from that point connection is plain HTTP.
2) Nginx passes info about used schema to Community Server so it could form HTTPS URLs
3) Browser doesn't speak with Document Server directly, but rather through Community Server as a kind of proxy.
So the connection looks like this
Browser <===> Nginx <---> CS <---> DS
Where === repesents HTTPS and --- represents HTTP

My wild guess is: since nginx passed to CS header that specifies HTTPS schema, CS in it's turn tries to do the same thing DS. At least i've seen 101 (Connection Upgrade) response for some of initial requests when trying to open document for editing.

I can be completely wrong, since I'm not an expert in this field, but if I'm right there should be a way to decouple schemas used by CS for external and internal communications.

**Comment by TMHBOFH** [comment]
@Zarbis 
thanks for your sharing knowledge.
I will search for other possibilities.

**Comment by renekalk** [comment]
anything new here? I have exact the same problem

**Comment by kuyper** [comment]
@smandolare, could you please give more detail on your solution? What exactly do you mean by "expose the document server container http port and give it a unique public name"? Did you define some kind of virtual host for the docker instance running the document server?

**Comment by smandolare** [comment]
If you installed OnlyOffice using the install script your docker images will be created with a docker network called something similar to onlyoffice-network and only the container for the community server has a host exposed port. The document server image will be part of the onlyoffice-network and be privately communicating with the community server over the docker network only.

to solve this issue I manually setup the docker image rather than using the script:
#Community Server
#Expose docker host port 80/443 (or whatever you want, this is the proxy destination port) mapped to container port 80 and 443
docker run -dit --restart=always -v /app/onlyoffice/CommunityServer/logs:/var/log/onlyoffice -v /app/onlyoffice/CommunityServer/data:/var/www/onlyoffice/Data -v /app/onlyoffice/CommunityServer/mysql:/var/lib/mysql --name docs-portal -p 80:80 -p 443:443 -p 5222:5222 --network onlyoffice onlyoffice/communityserver

#Document Server
#Expose docker host port 88 and 4433 similar to community server mapped to container ports 80 and 443
docker run -dit --name docs-editor --network onlyoffice -p 88:80 -p 4433:433 -v /app/onlyoffice/DocumentServer/data:/var/www/onlyoffice/Data -v /app/onlyoffice/DocumentServer/logs:/var/log/onlyoffice --restart=always onlyoffice/documentserver

On the reverse proxy I setup 2 virtual hosts which proxy to the appropriate docker port
#Community Portal
#Define SSL commands if doing HTTPS offloading
<VirtualHost *:80>
        #Public Server DNS Name
        ServerName docs.example.com
        RewriteEngine On
        RewriteCond %{HTTP:Upgrade} =websocket [NC]
        RewriteRule /(.*)           ws://10.99.0.200/$1 [P]
        ProxyPass / http://10.99.0.200/
        ProxyPassReverse https://docs.example.com/ http://10.99.0.200/
</VirtualHost>

#Document Server
#Define SSL commands if doing HTTPS offloading
<VirtualHost *:80>
        #Public Server DNS Name
        ServerName docs-editor.example.com
        RewriteEngine On
        RewriteCond %{HTTP:Upgrade} =websocket [NC]
        RewriteRule /(.*)           ws://10.99.0.200:88/$1 [P]
       #Notice the Port 88 to the document server defined above
        ProxyPass / http://10.99.0.200:88/
        ProxyPassReverse https://docs-editor.example.com/ http://10.99.0.200:88/
</VirtualHost>

# Since AWS offers free SSL certificates if used with their Elastic Load Balancer that is where my SSL offloading occurs and is not part of my reverse proxy config shown above.

![image](https://cloud.githubusercontent.com/assets/1140832/22046381/392e7c40-dcee-11e6-9391-bb5574a1ed48.png)

**Comment by kuyper** [comment]
Thank you! I'll try to create a comparable setup with nginx (which is my reverse proxy and does the SSL handling).

**Comment by smandolare** [comment]
I forgot to mention that this setup uses the document editor's full public https://URL in the portal integration page.

**Comment by kykc** [comment]
@smandolare thanks for the idea, this actually works!

So, if anyone is still interested in setting up such a setup with nginx, here's my success story:

docker-compose:

```yml
version: '2'

services:
  automatl-onlyoffice-cs:
    image: onlyoffice/communityserver:8.9.2.27
    restart: always
    stdin_open: true
    tty: true
    networks: 
      - onlyoffice
    container_name: automatl-onlyoffice-cs
    links:
     - automatl-onlyoffice-ds:document_server
    ports:
     - "172.17.0.1:5080:80"
     - "172.17.0.1:50443:443"
     - "5222:5222"
    volumes:
     - /home/user/onlyoffice/cs/logs:/var/log/onlyoffice
     - /home/user/onlyoffice/cs/data:/var/www/onlyoffice/Data
     - /home/user/onlyoffice/cs/mysql:/var/lib/mysql
     - /home/user/onlyoffice/ds/data:/var/www/onlyoffice/DocumentServerData
    environment:
     - MYSQL_SERVER_HOST=172.17.0.1
     - MYSQL_SERVER_DB_NAME=onlyoffice_cs
     - MYSQL_SERVER_USER=root
     - MYSQL_SERVER_PASS=secret_db
     - MYSQL_SERVER_PORT=3306
     - DOCUMENT_SERVER_PORT_80_TCP_ADDR=oods.example.org
  automatl-onlyoffice-ds:
    image: onlyoffice/documentserver:4.2.9.2
    restart: always
    stdin_open: true
    tty: true
    ports:
     - "172.17.0.1:5180:80"
    networks:
      - onlyoffice
    container_name: automatl-onlyoffice-ds
    volumes:
      - /home/user/onlyoffice/ds/data:/var/www/onlyoffice/Data
      - /home/user/onlyoffice/ds/logs:/var/log/onlyoffice

networks:
  onlyoffice:
    driver: bridge
```

Nginx:

```nginx
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}

server {
    listen 80;
    server_name office.example.org;
    return 301 https://$server_name$request_uri;
}

server {
    listen 80;
    server_name oods.example.org;
    return 301 https://$server_name$request_uri;
}

server {
    listen 443 ssl;
    ssl on;
    server_name oods.example.org;
    ssl_certificate /certs/chain.crt;
    ssl_certificate_key /certs/private.key;

    location /favicon.ico {
        return 204;
        access_log off;
        log_not_found off;
    }

    location / {
        proxy_pass http://172.17.0.1:5180;
        proxy_read_timeout 1800;
        proxy_connect_timeout 1800;
        proxy_set_header Host $http_host;
        proxy_redirect http://oods.example.org/ https://oods.example.org/;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}

server {
    listen 443 ssl;
    ssl on;
    server_name office.example.org;
    ssl_certificate /certs/chain.crt;
    ssl_certificate_key /certs/private.key;

    location = /favicon.ico {
        return 204;
        access_log off;
        log_not_found off;
    }

    location / {
	proxy_pass http://172.17.0.1:5080;
	proxy_read_timeout 1800;
	proxy_connect_timeout 1800;
	proxy_set_header Host $http_host;
	proxy_redirect http://office.example.org/ https://office.example.org/;
	proxy_set_header Upgrade $http_upgrade;
	proxy_set_header Connection $connection_upgrade;
	proxy_set_header X-Forwarded-Proto $scheme;
    }
}
```

So I have two containers on one host and nginx reverse proxy before them with two virtualhosts. It forwards queries from ``office.example.org`` to community server and queries for ``oods.example.org`` to document server.

After installation you may probably need to go to ``Settings -> Integration -> Document Service`` so it will look like this:

![screen](http://i.imgur.com/c1h93bn.png)

(note the ``https`` before document server endpoints)

**Comment by traedamatic** [comment]
Hello,
are there any news about a official approach for this issue? I had the same problem but i could almost solve it with the solution from @kykc. Unfortunately I get this new error:
`
2017-04-15 19:29:19,093 DEBUG [Threadpool worker] ASC.Files - DocService command Info fileId '8' docKey 'pqJRhCzsuX8lqjopicM_' callbackUrl 'https://office.XXXXX.XXX:443/products/files/httphandlers/filehandler.ashx?action=track&vkey=cmVHaHdxWWE4UEo1cEplTi9kemJhczB6OWFpMkhtTEYvWWdlaW9pamg3dz0_Ijgi0&new=false' users '' status ''
2017-04-15 19:29:19,137 ERROR [Threadpool worker] ASC.Files - DocService command error
System.Net.WebException: Error: SecureChannelFailure (The authentication or decryption has failed.) ---> System.IO.IOException: The authentication or decryption has failed. ---> System.IO.IOException: Error while sending TLS Alert (Fatal:InternalError): System.IO.IOException: The authentication or decryption has failed. ---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer. ---> System.Net.Sockets.SocketException: Connection reset by peer
  at System.Net.Sockets.Socket.EndReceive (System.IAsyncResult result) [0x00033] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.Sockets.NetworkStream.EndRead (System.IAsyncResult asyncResult) [0x0005f] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.RecordProtocol.EndReceiveRecord (System.IAsyncResult asyncResult) [0x00040] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.SslClientStream.SafeEndReceiveRecord (System.IAsyncResult ar, System.Boolean ignoreEmpty) [0x00000] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.SslClientStream.NegotiateAsyncWorker (System.IAsyncResult result) [0x00071] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslClientStream.EndNegotiateHandshake (System.IAsyncResult result) [0x0003b] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.SslStreamBase.AsyncHandshakeCallback (System.IAsyncResult asyncResult) [0x0000c] in <1d0bb82c94e7435eb09324cf5ef20e36>:0  ---> System.IO.IOException: Unable to write data to the transport connection: The socket is not connected. ---> System.Net.Sockets.SocketException: The socket is not connected
  at System.Net.Sockets.Socket.BeginSend (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.Net.Sockets.SocketFlags socket_flags, System.AsyncCallback callback, System.Object state) [0x0002b] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.Sockets.NetworkStream.BeginWrite (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.AsyncCallback callback, System.Object state) [0x000b4] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
   --- End of inner exception stack trace ---
  at System.Net.Sockets.NetworkStream.BeginWrite (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.AsyncCallback callback, System.Object state) [0x00113] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at Mono.Security.Protocol.Tls.RecordProtocol.BeginSendRecord (Mono.Security.Protocol.Tls.ContentType contentType, System.Byte[] recordData, System.AsyncCallback callback, System.Object state) [0x00026] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.RecordProtocol.SendRecord (Mono.Security.Protocol.Tls.ContentType contentType, System.Byte[] recordData) [0x00000] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.RecordProtocol.SendAlert (Mono.Security.Protocol.Tls.Alert alert) [0x00027] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Security.Protocol.Tls.RecordProtocol.SendAlert (System.Exception& ex) [0x00021] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslStreamBase.EndRead (System.IAsyncResult asyncResult) [0x00057] in <1d0bb82c94e7435eb09324cf5ef20e36>:0 
  at Mono.Net.Security.Private.LegacySslStream.EndAuthenticateAsClient (System.IAsyncResult asyncResult) [0x00011] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at Mono.Net.Security.Private.LegacySslStream.AuthenticateAsClient (System.String targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, System.Boolean checkCertificateRevocation) [0x0000e] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at Mono.Net.Security.MonoTlsStream.CreateStream (System.Byte[] buffer) [0x0006a] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
   --- End of inner exception stack trace ---
  at System.Net.HttpWebRequest.EndGetResponse (System.IAsyncResult asyncResult) [0x00064] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at System.Net.HttpWebRequest.GetResponse () [0x0000e] in <5071a6e4a4564e19a2eda0f53e42f9bd>:0 
  at ASC.Web.Core.Files.DocumentService.CommandRequest (System.String documentTrackerUrl, ASC.Web.Core.Files.DocumentService+CommandMethod method, System.String documentRevisionId, System.String callbackUrl, System.String users, System.String status, System.String& version) [0x000cb] in <af7252d722be4f26aa868d67d83eef1e>:0 
  at ASC.Web.Files.Services.DocumentService.DocumentServiceConnector.Command (ASC.Web.Core.Files.DocumentService+CommandMethod method, System.String docKeyForTrack, System.Object fileId, System.String callbackUrl, System.String users, System.String status) [0x00045] in <d3cce4f203504f01ab975d5ed96a63a8>:0 
2017-04-15 19:29:19,145 ERROR [Threadpool worker] ASC.Files - System.Exception: Fehler beim Starten der Dateibearbeitung
  at ASC.Web.Files.Services.WCFService.FileStorageServiceController.StartEdit (System.String fileId, System.String docKeyForTrack, System.Boolean asNew, System.Boolean editingAlone, System.String doc) [0x00057] in <d3cce4f203504f01ab975d5ed96a63a8>:0 `

The error is reported. See http://dev.onlyoffice.org/viewtopic.php?f=23&t=6858 or http://dev.onlyoffice.org/ru/viewtopic.php?f=30&t=1481 but all solutions don't work with my setup.

Did anyone have this error at some point?

Nginx Conf: 
[nginx-onlyoffice.txt](https://github.com/ONLYOFFICE/Docker-CommunityServer/files/923859/nginx-onlyoffice.txt)

Docker ps Output

`dbb59299a818        onlyoffice/communityserver:latest    "/bin/sh -c 'exec ..."   9 hours ago         Up 9 hours          443/tcp, 0.0.0.0:3306->3306/tcp, 0.0.0.0:9865-9866->9865-9866/tcp, 5280/tcp, 0.0.0.0:9871->9871/tcp, 9882/tcp, 0.0.0.0:9888->9888/tcp, 0.0.0.0:4003->80/tcp, 0.0.0.0:4004->5222/tcp   onlyoffice-community-server`

`8089709349cc        onlyoffice/documentserver:latest     "/bin/sh -c 'bash ..."   35 hours ago        Up 35 hours         443/tcp, 0.0.0.0:4005->80/tcp                                                                                                                                                         onlyoffice-document-server`

Thanks for any help,
Nicolas

**Comment by cubehouse** [comment]
Also getting this.

The error is around trying to load Editor.bin:
http://[MY DOMAIN]/cache/files/u5e_54ywql0_kgB_OGk_/Editor.bin/Editor.bin?md5=Eoin5nnsL1FusisreRcp5Q==&expires=1498765624&disposition=attachment&ooname=output.bin

Because the app is loaded over https, this fails. How can I tell the application to load this with https?
The URL to load seems to come from a websocket response, which is making it very hard to track down...

**Comment by cubehouse** [comment]
Created a workaround!

I copied all the files from here: https://github.com/ONLYOFFICE/server/tree/master/Common/config
and placed them into a local directory.

I then mounted this directory to my document-server docker container at /etc/onlyoffice/documentserver/

Then, I modified my local production-linux.json to add storage.externalHost to match my domain (so it read "https://MYDOMAIN/".

Now when it goes to load Editor.bin, it uses externalHost as the base URL. There are a ton more errors with other things loading over http now, but the editor generally seems to almost work now, so progress I guess!......

**Comment by crazybadger** [comment]
Useful discussion and thanks @kykc for the nginx.conf

I've just discovered Onlyoffice CE, although have been using Onlyoffice Document Server (DS) for a while with ownCloud.  I was previously spinning up a ubuntu VM to run Onlyoffice DS on my FreeNAS machine and was trying to replicate this using Docker.  I accidentally created a container for CE, so have been playing :-)

All this is running behind a Nginx reverse proxy, so the file helped me redirecting to the different contains I have running CE and DS.  Its working to a point, and I can still edit documents from ownCloud and the DS server starts when I edit a document in CE, but when I start to edit the document I get a 'File editing start error'.  Also can't add new documents, but this is less of an issue as I have added a connection to my ownCloud server, so will be mainly adding documents through the desktop client sync.

I also create a Nextcloud container to see how that's forked away from ownCloud, but with pretty much the same reverse proxy settings I can't even get that to link to the Onlyoffice DS.

Will keep playing, although if anyone has any suggestions that would be great!
